jbmc/src/java_bytecode/README.md

```diff
@@ -43,6 +43,45 @@ To be documented.
 
 To be documented.
 
+\section java-bytecode-remove-instanceof Remove instanceof
+
+Java has an operator called `instanceof`, which is used to check if an instance of a class is of a particular type. For example given that `A` and `B` are the identifiers of classes, which derive directly from Object. If the following example java code is executed -
+```java
+Object a = new A();
+boolean x = a instanceof A; // true
+boolean y = a instanceof B; // false
+```
+Then, `x` will be set to `true` and `y` will be set to `false`.
+
+`instanceof` also works with inheritance. Given classes `C` and `D`, where `C` derrives from `Object` and `D` derives from `C`. If the following example java code is executed -
+```java
+Object c = new C();
+Object d = new D();
+boolean b1 = a instanceof A; // true
+boolean b2 = a instanceof B; // false
+boolean b3 = b instanceof A; // true
+boolean b4 = b instanceof B; // true
+```
+Then, `b1` will be `true`, `b2` will be `false`, `b3` will be `true` and `b4` will be `true`.
+
+`instanceof` has its own instruction type in the java bytecode. We need to rewrite it in terms of other, more fundamental operations in order to analyse it. This operation is referred to as "lowering" or in this specific case - `remove_instanceof`.
+
+For a simple example where class `A` is derived from `Object` and no classes derrive from `A`, the expression `e instanceof A` needs to be lowered to an expression equivalent to `e == null ? false : e->@class_identifier == "A"`. The null case needs to be taken into account first, as we can't lookup the value of a `@class_identifier` field in a null pointer. Note that according to the Java specification, `null` is not an instance of any type.
+
+For an example with inheritance where class `D` is derrived from `C` and `C` is derrived from `Object`, then `e instanceof C` needs to be lowered to an expression equivalent to `e == null ? false : (e->@class_identifier == "C" || e->@class_identifier == "D")`. To lower `e instanceof C` correctly, we need to take into account each of the classes derrived from `C`. Working out what classes derive from a given class type is performed using an instance of the `class_hierarchyt` class.
```

Comment:
I would do one example (the slightly more complex one) and work through that since the info in the first one is entirely contained in the second and it isn't super complex IMO.